File: 백준/Gold/1339. 단어 수학/단어 수학.js
const fs = require("fs");

let read = fs.readFileSync("dev/stdin").toString().trim();

let [n, ...arr] = read.split("\n");
arr = arr.map((s) => [...s.trim()]);
const mapper = {};

const setMapper = () => {
  const counter = {};
  let maxv = Math.max(...arr.map((s) => s.length));
  for (let i = -1; i >= -maxv; i--) {
    for (let a of arr) {
      if (!a.at(i)) continue;
      counter[a.at(i)] = (counter[a.at(i)] ?? 0) + 10 ** (i * -1 - 1);
    }
  }

  const sorted = Object.keys(counter).sort((a, b) => counter[b] - counter[a]);

  for (let i = 9; i > 9 - sorted.length; i--) {
    mapper[sorted[9 - i]] = i;
  }
};

const main = () => {
  setMapper();
  return arr.reduce(
    (acc, a) => acc + parseInt(a.map((s) => mapper[s]).join("")),
    0
  );
};

console.log(main());

File: 백준/Gold/1339. 단어 수학/단어 수학.py
import sys

input = sys.stdin.readline

n = int(input())
arr = [input().strip() for _ in range(n)]
mapper = {}


def setMapper():
    global mapper
    dic = {}
    for i in range(-1, -max([len(a) for a in arr]) - 1, -1):
        for a in arr:
            if i < -len(a):
                continue
            dic[a[i]] = dic.get(a[i], 0) + (10 ** ((i * -1) - 1))

    sort_li = sorted(dic.keys(), key=lambda x: dic[x], reverse=True)

    for i in range(9, 9 - len(sort_li), -1):
        mapper[sort_li[9 - i]] = i


def main():
    setMapper()
    sumv = 0
    for a in arr:
        sumv += int("".join([str(mapper[s]) for s in a]))

    return sumv


print(main())
